File: src/main/scala/string/LongestSubstringWithoutRepeatingCharacters.scala
package String

import scala.collection.mutable

object LongestSubstringWithoutRepeatingCharacters {
  def lengthOfLongestSubstring(s: String): Int = {
    var i = 0
    var j = 0
    var max = 0
    val n = s.length
    val set = new mutable.HashSet[Char]()
    while(i < n && j < n){
      if(!set.contains(s.charAt(j))){
        set.+=(s.charAt(j))
        max = math.max(max,j - i + 1)
        j += 1
      }else{
        set -= s.charAt(i)
        i += 1
      }
    }
    max
  }
  def lengthOfLongestSubstring1(s: String): Int = {
    val map = new mutable.HashMap[Char,Int]()
    var j = 0
    var max = 0
    for(i <- s.indices){
      if(map.contains(s.charAt(i))){
        j = math.max(map(s.charAt(i)),j)
      }
      max = math.max(max,i - j + 1)
      map += s.charAt(i) -> (i + 1)
    }
    max
  }
}

File: src/main/scala/string/LongestSubstringWithoutRepeatingCharacters.java
package string;

import java.util.HashSet;
import java.util.Set;

/**
 * @Author: xck
 * @File: LongestSubstringWithoutRepeatingCharacters
 * @Time: 14:28 2019/12/3
 */
public class LongestSubstringWithoutRepeatingCharacters {
    public int lengthOfLongestSubstring(String s) {

        int max = 0;

        StringBuffer buffer = new StringBuffer();
        for(int i = 0;i < s.length();i++){
            int count = 0;
            Set<Character> set = new HashSet<>();
            for(int j = i ;j < s.length();j++){
                if(set.add(s.charAt(j))){
                    count++;
                }else
                    break;
            }
            max = Math.max(max,count);
        }
        return max;
    }

    public int lengthOfLongestSubstring1(String s) {

        int max = 0;
        Set<Character> set = new HashSet<>();
        int j = 0;
        for(int i = 0;i < s.length();){
            if(!set.contains( s.charAt(i))){
                set.add(s.charAt(i));
                i++;
            }else{
                set.remove( s.charAt(j++));
            }
            max = Math.max(set.size(),max);
        }
        return max;
    }
}
